In [17]:
import numpy as np
import pandas as pd 
from pandas import DataFrame, Series
import matplotlib.pyplot as plt

In [18]:
# read data into new pandas DataFrame.
data = pd.read_csv('complete_players_data.csv')

/home/yarko/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (27,39) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [19]:
data

ID               Name  Age  \
0       20801  Cristiano Ronaldo   32   
1      158023           L. Messi   30   
2      190871             Neymar   25   
3      176580          L. Suárez   30   
4      167495           M. Neuer   31   
5      188545     R. Lewandowski   28   
6      193080             De Gea   26   
7      183277          E. Hazard   26   
8      182521           T. Kroos   27   
9      167664         G. Higuaín   29   
10     155862       Sergio Ramos   31   
11     192985       K. De Bruyne   26   
12     192119        T. Courtois   25   
13     184941         A. Sánchez   28   
14     177003          L. Modrić   31   
15     173731            G. Bale   27   
16     153079          S. Agüero   29   
17     138956       G. Chiellini   32   
18       1179          G. Buffon   39   
19     211110          P. Dybala   23   
20     200389           J. Oblak   24   
21     194765       A. Griezmann   26   
22     189509             Thiago   26   
23     188567      P. Aubameyang   28   
24     184344         L. Bonucci   30   
25     183907         J. Boateng   28   
26     182493           D. Godín   31   
27     178603         M. Hummels   28   
28     176635            M. Özil   28   
29     167948          H. Lloris   30   
...       ...                ...  ...   
17951  221669           M. Hurst   21   
17952  239079       A. Osmanoski   17   
17953  238833          K. Cotter   18   
17954  225778        T. Robinson   19   
17955  231928          R. Hughes   18   
17956  240381        Z. Mohammed   17   
17957  236807         D. Peppard   17   
17958  237859          C. Rogers   17   
17959  225319      N. McLaughlin   18   
17960  236597           L. Kiely   18   
17961  233047   J. Latibeaudiere   17   
17962  236425           J. Payne   18   
17963  238217          G. Manley   19   
17964  238985        P. Phillips   18   
17965  238219           A. Byrne   18   
17966  238477        K. Fujikawa   18   
17967  231824            K. Egan   19   
17968  237974      T. Brownsword   17   
17969  236954         F. Prohart   18   
17970  231107         A. Kilgour   19   
17971  240325           R. White   18   
17972  238306          A. Conway   19   
17973  240403          T. Sawyer   18   
17974  240404          J. Keeble   18   
17975  235352       T. Käßemodel   28   
17976  237463          A. Kelsey   17   
17977   11728      B. Richardson   47   
17978  231381           J. Young   17   
17979  238813       J. Lundstram   18   
17980  238308          L. Sackey   18   

                                                 Photo          Nationality  \
0       https://cdn.sofifa.org/48/18/players/20801.png             Portugal   
1      https://cdn.sofifa.org/48/18/players/158023.png            Argentina   
2      https://cdn.sofifa.org/48/18/players/190871.png               Brazil   
3      https://cdn.sofifa.org/48/18/players/176580.png              Uruguay   
4      https://cdn.sofifa.org/48/18/players/167495.png              Germany   
5      https://cdn.sofifa.org/48/18/players/188545.png               Poland   
6      https://cdn.sofifa.org/48/18/players/193080.png                Spain   
7      https://cdn.sofifa.org/48/18/players/183277.png              Belgium   
8      https://cdn.sofifa.org/48/18/players/182521.png              Germany   
9      https://cdn.sofifa.org/48/18/players/167664.png            Argentina   
10     https://cdn.sofifa.org/48/18/players/155862.png                Spain   
11     https://cdn.sofifa.org/48/18/players/192985.png              Belgium   
12     https://cdn.sofifa.org/48/18/players/192119.png              Belgium   
13     https://cdn.sofifa.org/48/18/players/184941.png                Chile   
14     https://cdn.sofifa.org/48/18/players/177003.png              Croatia   
15     https://cdn.sofifa.org/48/18/players/173731.png                Wales   
16     https://cdn.sofifa.org/48/18/players/153079.png            Argentina   
17     https://cdn.sofifa.org/48/18/players/13

In [20]:
# initial number of rows
len(data.index)

17981

In [21]:
# check, whether there are any NaN values in the dataset.
data.isnull().values.any()

True

In [22]:
# count how many NaN values are there in every column (omitting 0 quantities).
null_columns=data.columns[data.isnull().any()]
data[null_columns].isnull().sum()

Club             248
League           248
League Level     248
Country          248
Region           248
CAM             2029
CB              2029
CDM             2029
CF              2029
CM              2029
LAM             2029
LB              2029
LCB             2029
LCM             2029
LDM             2029
LF              2029
LM              2029
LS              2029
LW              2029
LWB             2029
RAM             2029
RB              2029
RCB             2029
RCM             2029
RDM             2029
RF              2029
RM              2029
RS              2029
RW              2029
RWB             2029
ST              2029
dtype: int64

#### column *Club* has 248 NaN values, which means, that there are 248 free agents (players, currently without a club)

In [23]:
# print players, that are free agents.
print(data[data["Club"].isnull()]["Name"])

162                Oscar
167         Adrien Silva
274            A. Witsel
472              M. Berg
488       Renato Augusto
658             Gervinho
925         B. Moukandjo
944            B. Dočkal
1019                 Gil
1273          C. Riveros
1351           Zheng Zhi
1402              P. Tau
1549          E. Paredes
1597       T. Hlatshwayo
1802         D. González
1819          M. Martins
1826       J. Beausejour
1888        C. Rodríguez
1899          C. Toselli
2000             G. Jara
2090          Zeng Cheng
2241            T. Sivok
2246        T. Sainsbury
2281             Gao Lin
2350             Z. Gera
2424          P. Gallese
2483         V. Stoyanov
2518          J. Herrera
2552       Feng Xiaoting
2553       J. Fuenzalida
              ...       
14723      J. Lalpekhlua
14805     B. Makendzhiev
14817         Gao Zhunyi
14865    G. Singh Sandhu
14921        Deng Hanwen
15237         K. Pereira
15257           J. Singh
15330           B. Jairu
15564           G. Bokwé


In [24]:
# drop rows (by index), that contain free agents.
rows_to_drop = data[data["Club"].isnull()]["Name"].index.tolist();
data.drop(rows_to_drop,inplace=True)

In [25]:
# check, if done correctly.
data[null_columns].isnull().sum()

Club               0
League             0
League Level       0
Country            0
Region             0
CAM             1987
CB              1987
CDM             1987
CF              1987
CM              1987
LAM             1987
LB              1987
LCB             1987
LCM             1987
LDM             1987
LF              1987
LM              1987
LS              1987
LW              1987
LWB             1987
RAM             1987
RB              1987
RCB             1987
RCM             1987
RDM             1987
RF              1987
RM              1987
RS              1987
RW              1987
RWB             1987
ST              1987
dtype: int64

#### as we can see, all rows with free agents have been dropped.

In [26]:
# the number of rows is reduced accordingly.
len(data)

17733

In [27]:
# find players with empty name field (if any)
counter = 0;
for row in data["Name"]:
    if row == "":
        counter += 1;
counter

0

#### so, there aren't any players with empty string as a name in our dataset.

In [28]:
# print(data["Preferred Positions"])
# convert "Preferred Positions" column values to single position and remove whitespace at the end.
data["Preferred Positions"].replace(regex={r'^(\w+).*$':r'\1'}, inplace=True)

In [29]:
data["Preferred Positions"]

0         ST
1         RW
2         LW
3         ST
4         GK
5         ST
6         GK
7         LW
8        CDM
9         ST
10        CB
11        RM
12        GK
13        RM
14       CDM
15        RW
16        ST
17        CB
18        GK
19        ST
20        GK
21        LW
22       CDM
23        ST
24        CB
25        CB
26        CB
27        CB
28        RW
29        GK
        ... 
17951     GK
17952     CM
17953     RB
17954     LM
17955     CM
17956     CB
17957     RB
17958     CB
17959    CAM
17960    CAM
17961     CB
17962     CB
17963     CM
17964     CM
17965     CM
17966     CM
17967     RB
17968     CB
17969     ST
17970     CB
17971     ST
17972     CB
17973     CM
17974     CB
17975     CM
17976     GK
17977     GK
17978     ST
17979     CM
17980     ST
Name: Preferred Positions, Length: 17733, dtype: object

In [30]:
len(data["Preferred Positions"])

17733

In [34]:
# Check if 1987 NaN values for playing position columns are GK (goalkeepers), since they only play one position.
# So, count number of rows, for which the value of "Preferred Positions" column is "GK"
gk_match = data['Preferred Positions'] == "GK"
len(data[gk_match])

1987

#### so, the number of GKs is exactly 1987. Since they don't play any other position, NaN values they have in other positions columns need not be filled with something else. In fact, they are handled in a proper way by pandas methods.

In [36]:
# drop columns of the dataset, that are not useful for analysis purposes
data.drop(columns = ['Photo','Flag','Club Logo','Special'], inplace=True)

In [59]:
print(data.columns)
[data[col].dtype for col in data.columns]

Index(['ID', 'Name', 'Age', 'Nationality', 'Overall', 'Potential', 'Club',
       'League', 'League Level', 'Country', 'Region', 'Value', 'Wage',
       'Acceleration', 'Aggression', 'Agility', 'Balance', 'Ball control',
       'Composure', 'Crossing', 'Curve', 'Dribbling', 'Finishing',
       'Free kick accuracy', 'GK diving', 'GK handling', 'GK kicking',
       'GK positioning', 'GK reflexes', 'Heading accuracy', 'Interceptions',
       'Jumping', 'Long passing', 'Long shots', 'Marking', 'Penalties',
       'Positioning', 'Reactions', 'Short passing', 'Shot power',
       'Sliding tackle', 'Sprint speed', 'Stamina', 'Standing tackle',
       'Strength', 'Vision', 'Volleys', 'Preferred Positions', 'CAM', 'CB',
       'CDM', 'CF', 'CM', 'LAM', 'LB', 'LCB', 'LCM', 'LDM', 'LF', 'LM', 'LS',
       'LW', 'LWB', 'RAM', 'RB', 'RCB', 'RCM', 'RDM', 'RF', 'RM', 'RS', 'RW',
       'RWB', 'ST'],
      dtype='object')


[dtype('int64'),
 dtype('O'),
 dtype('int64'),
 dtype('O'),
 dtype('int64'),
 dtype('int64'),
 dtype('O'),
 dtype('O'),
 dtype('float64'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('O'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('float64'),
 dtype('floa

In [69]:
# change dtype of "League Level" to integer.
data["League Level"] = data["League Level"].astype(int)

In [76]:
# convert wage and value column values to numbers.
def convert_currency(curr_col):
    without_euro_symbol = curr_col.str[1:]
    unit_symbol = without_euro_symbol.str[-1]
    numeric_part = np.where(unit_symbol == '0', 0, without_euro_symbol.str[:-1].pipe(pd.to_numeric))
    multipliers = unit_symbol.replace({'M': 1e6, 'K': 1e3}).pipe(pd.to_numeric)
    return numeric_part * multipliers

In [82]:
# "Wage" column is now a floating point number.
data["Wage"] = convert_currency(data["Wage"])

In [84]:
# "Value" column is now a floating point number.
data["Value"] = convert_currency(data["Value"])

In [92]:
# columns Acceleration ... Volleys have dtype object. We will convert them to float.
pd.DataFrame({"Column":data.columns,"DType":[data[col].dtype for col in data.columns]})

Column    DType
0                    ID    int64
1                  Name   object
2                   Age    int64
3           Nationality   object
4               Overall    int64
5             Potential    int64
6                  Club   object
7                League   object
8          League Level    int64
9               Country   object
10               Region   object
11                Value  float64
12                 Wage  float64
13         Acceleration   object
14           Aggression   object
15              Agility   object
16              Balance   object
17         Ball control   object
18            Composure   object
19             Crossing   object
20                Curve   object
21            Dribbling   object
22            Finishing   object
23   Free kick accuracy   object
24            GK diving   object
25          GK handling   object
26           GK kicking   object
27       GK positioning   object
28          GK reflexes   object
29     Heading accuracy   object
..                  ...      ...
44             Strength   object
45               Vision   object
46              Volleys   object
47  Preferred Positions   object
48                  CAM  float64
49                   CB  float64
50                  CDM  float64
51                   CF  float64
52                   CM  float64
53                  LAM  float64
54                   LB  float64
55                  LCB  float64
56                  LCM  float64
57                  LDM  float64
58                   LF  float64
59                   LM  float64
60                   LS  float64
61                   LW  float64
62                  LWB  float64
63                  RAM  float64
64                   RB  float64
65                  RCB  float64
66                  RCM  float64
67                  RDM  float64
68                   RF  float64
69                   RM  float64
70                   RS  float64
71                   RW  float64
72                  RWB  float64
73                   ST  float64

[74 rows x 2 columns]

In [93]:
# Example of wrong data in mentioned above columns (Dribbling = 78+3 => 78)
data[data["ID"]==189513].T

157
ID                                      189513
Name                                    Parejo
Age                                         28
Nationality                              Spain
Overall                                     83
Potential                                   83
Club                               Valencia CF
League               Spanish Primera División 
League Level                                 1
Country                                  Spain
Region                                  Europe
Value                                 2.55e+07
Wage                                     45000
Acceleration                                55
Aggression                               58-10
Agility                                     69
Balance                                     72
Ball control                                83
Composure                                   74
Crossing                                    84
Curve                                       87
Dribbling                                 78+3
Finishing                                   75
Free kick accuracy                          90
GK diving                                    5
GK handling                                  9
GK kicking                                  12
GK positioning                              14
GK reflexes                                  8
Heading accuracy                            63
...                                        ...
Strength                                    72
Vision                                      85
Volleys                                     72
Preferred Positions                        CDM
CAM                                         80
CB                                          65
CDM                                         74
CF                                          78
CM                                          82
LAM                                         80
LB                                          68
LCB                                         65
LCM                                         82
LDM                                         74
LF                                          78
LM                                          78
LS                                          75
LW                                          77
LWB                                         71
RAM                                         80
RB                                          68
RCB                                         65
RCM                                         82
RDM                                         74
RF                                          78
RM                                          78
RS                                          75
RW                                          77
RWB                                         71
ST                                          75

[74 rows x 1 columns]

In [95]:
# Correct attributes representation and convert dtype of attribute columns (Acceleration ... Volleys) to integer.
for col in data.columns[13:47]:
    data[col].replace(regex={r'^(\d+)[\+\-].*$':r'\1'}, inplace=True)
    data[col] = pd.to_numeric(data[col])
pd.DataFrame({"Column":data.columns,"DType":[data[col].dtype for col in data.columns]})

Column    DType
0                    ID    int64
1                  Name   object
2                   Age    int64
3           Nationality   object
4               Overall    int64
5             Potential    int64
6                  Club   object
7                League   object
8          League Level    int64
9               Country   object
10               Region   object
11                Value  float64
12                 Wage  float64
13         Acceleration    int64
14           Aggression    int64
15              Agility    int64
16              Balance    int64
17         Ball control    int64
18            Composure    int64
19             Crossing    int64
20                Curve    int64
21            Dribbling    int64
22            Finishing    int64
23   Free kick accuracy    int64
24            GK diving    int64
25          GK handling    int64
26           GK kicking    int64
27       GK positioning    int64
28          GK reflexes    int64
29     Heading accuracy    int64
..                  ...      ...
44             Strength    int64
45               Vision    int64
46              Volleys    int64
47  Preferred Positions   object
48                  CAM  float64
49                   CB  float64
50                  CDM  float64
51                   CF  float64
52                   CM  float64
53                  LAM  float64
54                   LB  float64
55                  LCB  float64
56                  LCM  float64
57                  LDM  float64
58                   LF  float64
59                   LM  float64
60                   LS  float64
61                   LW  float64
62                  LWB  float64
63                  RAM  float64
64                   RB  float64
65                  RCB  float64
66                  RCM  float64
67                  RDM  float64
68                   RF  float64
69                   RM  float64
70                   RS  float64
71                   RW  float64
72                  RWB  float64
73                   ST  float64

[74 rows x 2 columns]

In [116]:
# check if ID is unique, if so, set it as an index
data["ID"].is_unique

False

#### so, we now delete the duplicates

In [117]:
data.drop_duplicates('ID', keep='first', inplace=True)

In [122]:
# confirm, there are no duplicates
print(data[data["ID"].duplicated()])
data["ID"].is_unique

Empty DataFrame
Columns: [ID, Name, Age, Nationality, Overall, Potential, Club, League, League Level, Country, Region, Value, Wage, Acceleration, Aggression, Agility, Balance, Ball control, Composure, Crossing, Curve, Dribbling, Finishing, Free kick accuracy, GK diving, GK handling, GK kicking, GK positioning, GK reflexes, Heading accuracy, Interceptions, Jumping, Long passing, Long shots, Marking, Penalties, Positioning, Reactions, Short passing, Shot power, Sliding tackle, Sprint speed, Stamina, Standing tackle, Strength, Vision, Volleys, Preferred Positions, CAM, CB, CDM, CF, CM, LAM, LB, LCB, LCM, LDM, LF, LM, LS, LW, LWB, RAM, RB, RCB, RCM, RDM, RF, RM, RS, RW, RWB, ST]
Index: []

[0 rows x 74 columns]


True

In [123]:
# check column "ID" as an index
data.set_index("ID")

Name  Age          Nationality  Overall  Potential  \
ID                                                                        
20801   Cristiano Ronaldo   32             Portugal       94         94   
158023           L. Messi   30            Argentina       93         93   
190871             Neymar   25               Brazil       92         94   
176580          L. Suárez   30              Uruguay       92         92   
167495           M. Neuer   31              Germany       92         92   
188545     R. Lewandowski   28               Poland       91         91   
193080             De Gea   26                Spain       90         92   
183277          E. Hazard   26              Belgium       90         91   
182521           T. Kroos   27              Germany       90         90   
167664         G. Higuaín   29            Argentina       90         90   
155862       Sergio Ramos   31                Spain       90         90   
192985       K. De Bruyne   26              Belgium       89         92   
192119        T. Courtois   25              Belgium       89         92   
184941         A. Sánchez   28                Chile       89         89   
177003          L. Modrić   31              Croatia       89         89   
173731            G. Bale   27                Wales       89         89   
153079          S. Agüero   29            Argentina       89         89   
138956       G. Chiellini   32                Italy       89         89   
1179            G. Buffon   39                Italy       89         89   
211110          P. Dybala   23            Argentina       88         93   
200389           J. Oblak   24             Slovenia       88         93   
194765       A. Griezmann   26               France       88         91   
189509             Thiago   26                Spain       88         90   
188567      P. Aubameyang   28                Gabon       88         88   
184344         L. Bonucci   30                Italy       88         88   
183907         J. Boateng   28              Germany       88         88   
182493           D. Godín   31              Uruguay       88         88   
178603         M. Hummels   28              Germany       88         88   
176635            M. Özil   28              Germany       88         88   
167948          H. Lloris   30               France       88         88   
...                   ...  ...                  ...      ...        ...   
221669           M. Hurst   21             Scotland       48         58   
239079       A. Osmanoski   17        FYR Macedonia       48         68   
238833          K. Cotter   18              England       48         67   
225778        T. Robinson   19              England       48         64   
231928          R. Hughes   18             Scotland       48         63   
240381        Z. Mohammed   17              England       48         64   
236807         D. Peppard   17  Republic of Ireland       47         61   
237859          C. Rogers   17  Republic of Ireland       47         59   
225319      N. McLaughlin   18             Scotland       47         64   
236597           L. Kiely   18  Republic of Ireland       47         69   
233047   J. Latibeaudiere   17              England       47         73   
236425           J. Payne   18              England       47         63   
238217          G. Manley   19  Republic of Ireland       47         58   
238985        P. Phillips   18  Republic of Ireland       47         67   
238219           A. Byrne   18  Republic of Ireland       47         61   
238477        K. Fujikawa   18                Japan       47         67   
231824            K. Egan   19              England       47         67   
237974      T. Brownsword   17              England       47         68   
236954         F. Prohart   18              Austria       47         67   
231107         A. Kilgour   19              England       47         56   
240325           R. White   18              En

In [124]:
# confirm data integrity
data[null_columns].isnull().sum()

Club               0
League             0
League Level       0
Country            0
Region             0
CAM             1982
CB              1982
CDM             1982
CF              1982
CM              1982
LAM             1982
LB              1982
LCB             1982
LCM             1982
LDM             1982
LF              1982
LM              1982
LS              1982
LW              1982
LWB             1982
RAM             1982
RB              1982
RCB             1982
RCM             1982
RDM             1982
RF              1982
RM              1982
RS              1982
RW              1982
RWB             1982
ST              1982
dtype: int64

#### so, data integrity is confirmed (number of GK's reduced, courtesy of duplicates)

In [125]:
data.to_csv('cleaned_data.csv', encoding='utf-8')